[WebDriver][WPE] Support creating new webviews from create-web-view signal

lauromoura opened the pull request:
#### 474a4428f7ed3cc1525d935f010d1d0979c52c9f
<pre>
[WebDriver][WPE] Support creating new webviews from create-web-view signal
<a href="https://bugs.webkit.org/show_bug.cgi?id=212950">https://bugs.webkit.org/show_bug.cgi?id=212950</a>

Reviewed by Carlos Garcia Campos.

The create-web-view signal is used by the automation session to ask the
user code to return new views for automation. For example, the first
view being automated or new views as response to WebDriver commands
creating new windows/tabs.

In the new WPE API, it&apos;s easier to create new WebViews. This commit
takes advantage of this to support this feature (currently only through
new windows, as MiniBrowser doesn&apos;t support tabs). In the old WPE API,
we keep the previous behavior of always returning the same view.

* Tools/MiniBrowser/wpe/main.cpp:
(createWebViewForAutomationCallback):

Canonical link: <a href="https://commits.webkit.org/280534@main">https://commits.webkit.org/280534@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0df1ab0f66abbd9716b7cadaac305b04f292774e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56803 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36131 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9277 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60419 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7246 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58931 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43754 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7436 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46006 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5073 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58833 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33944 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49022 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26863 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30724 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6353 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6250 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52656 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6624 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62101 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/716 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6728 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53265 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/719 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49077 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53287 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12578 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/606 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31960 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33045 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34130 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32791 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -177,9 +177,40 @@ static gboolean wpeViewEventCallback(WPEView* view, WPEEvent* event, WebKitWebVi
 #endif
 
+static WebKitWebView* createWebView(WebKitWebView*, WebKitNavigationAction*, gpointer user_data);
+static void webViewClose(WebKitWebView*, gpointer user_data);
```
Could you move `createWebViewForAutomationCallback` after these two?

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -177,9 +177,40 @@ static gboolean wpeViewEventCallback(WPEView* view, WPEEvent* event, WebKitWebVi
+        firstAutomationView = false;
+        g_print("Reusing initial view as the first automated view\n");
+        return view;
+    }
```
I think it's ok to reuse the view if it's the only one and nothing has been loaded yet, that's what we do in GTK port. So, you could check the openViews hash table, if size is 1 and the uri of the view is NULL then you can reuse it. I don't think we need to show a printf message for this.

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -177,9 +177,40 @@ static gboolean wpeViewEventCallback(WPEView* view, WPEEvent* event, WebKitWebVi
+        return view;
+    }
+
+    g_print("Creating new view for automation from related view %p\n", view);
```
Same here, let's not add noise to stdout.

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -177,9 +177,40 @@ static gboolean wpeViewEventCallback(WPEView* view, WPEEvent* event, WebKitWebVi
+
+    g_print("Creating new view for automation from related view %p\n", view);
+    auto* newWebView = WEBKIT_WEB_VIEW(g_object_new(WEBKIT_TYPE_WEB_VIEW,
+        "related-view", view,
```
Why related view? I don't think it should use the same web process. Pass the same context, and make sure `is-controlled-by-automation` is TRUE too.

carlosgcampos commented on Tools/MiniBrowser/wpe/main.cpp:
```diff
@@ -177,9 +177,40 @@ static gboolean wpeViewEventCallback(WPEView* view, WPEEvent* event, WebKitWebVi
+    g_signal_connect(newWebView, "close", G_CALLBACK(webViewClose), application);
+
+    g_hash_table_add(openViews, newWebView);
+
```
For this to work, we need to make sure we don't load anything (we currently load about:blank in activate()) first, and here we load about:blank.

carlosgcampos reviewed (approved):
Excellent!

lauromoura committed:
[WebDriver][WPE] Support creating new webviews from create-web-view signal
https://bugs.webkit.org/show_bug.cgi?id=212950

Reviewed by Carlos Garcia Campos.

The create-web-view signal is used by the automation session to ask the
user code to return new views for automation. For example, the first
view being automated or new views as response to WebDriver commands
creating new windows/tabs.

In the new WPE API, it's easier to create new WebViews. This commit
takes advantage of this to support this feature (currently only through
new windows, as MiniBrowser doesn't support tabs). In the old WPE API,
we keep the previous behavior of always returning the same view.

* Tools/MiniBrowser/wpe/main.cpp:
(createWebViewForAutomationCallback):

Canonical link: https://commits.webkit.org/280534@main
